src/goto-instrument/contracts/doc/user-guide/contracts-assigns.md

```diff
@@ -6,39 +6,137 @@
 
 ### Syntax
 
+An _assigns_ clause allows the user to specify a set of locations that may be
+assigned to by a function or the body of a loop:
+
 ```c
-__CPROVER_assigns(*identifier*, ...)
+__CPROVER_assigns(*targets*)
+```
+
+A function or loop contract contract may have zero or more _assigns_ clauses.
+
+The clause accepts a semicolon-separated list of _conditional target groups_.
+A _conditional target group+ consists of an optional _condition_ followed by a
+coma-separated list of _targets_.
+A _target_ is either an lvalue expression or a call to a function returning
+the built-in type `__CPROVER_assignable_t`.
+
+```
+targets ::= conditional_target_group (';' conditional_target_group)* (';')?
+conditional_target_group ::= (condition ':')? target (',' target)*
+target ::= lvalue-expression | call-to-cprover-assignable-t-function
 ```
 
-An _assigns_ clause allows the user to specify that a memory location may be written by a function. The set of locations writable by a function is the union of the locations specified by the assigns clauses, or the empty set of no _assigns_ clause is specified. While, in general, an _assigns_ clause could be interpreted with either _writes_ or _modifies_ semantics, this
-design is based on the former. This means that memory not captured by an
-_assigns_ clause must not be written within the given function, even if the
-value(s) therein are not modified.
+The set of locations writable by a function is the union of the locations
+specified by the assigns clauses, or the empty set of no _assigns_ clause is
+specified.
+While, in general, an _assigns_ clause could be interpreted with either
+_writes_ or _modifies_ semantics, this design is based on the former.
+This means that memory not captured by an _assigns_ clause must not be assigned
+to by the given function, even if the value(s) therein are not modified.
 
 ### Object slice expressions
 
-The following functions can be used in assigns clause to specify ranges of 
-assignable addresses.
+The following functions can be used in assigns clauses to specify ranges of assignable bytes.
+
+Given an lvalue expression `expr` with a complete type `expr_t`,
+ `__CPROVER_typed_target(expr)` specifies that the range
+ of `sizeof(expr_t)` bytes starting at `&expr` is assignable:
+```c
+__CPROVER_assignable_t __CPROVER_typed_target(expr_t expr);
+```
+
+Given a pointer `ptr` pointing into some object `o`,
+`__CPROVER_whole_object(ptr)` specifies that all bytes of the object `o`
+are assignable:
+```c
+__CPROVER_assignable_t __CPROVER_whole_object(void *ptr);
+```
+
+Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_from(ptr)`
+specifies that the range of bytes starting from the pointer and until the end of
+the object `o` are assignable:
+```c
+__CPROVER_assignable_t __CPROVER_object_from(void *ptr);
+```
+
+Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_upto(ptr, size)`
```

Comment:
Can we go back to the name "object_slice" instead of "object_upto".
Or, can we rename "havoc_slice" to "havoc_upto"? These things should have related names.